src/main.rs

```diff
@@ -26,10 +33,76 @@ fn main() {
     }
 }
 
+fn run_in_userns() -> anyhow::Result<()> {
+    // dnf needs to be run as root, but given that rpmoci only needs to query package repos
+    // and/or install packages into an install root, we can run in a user namespace, mapping
+    // the current uid/gid to root
+    // this function spawns a child process in a new user namespace, the parent configures
+    // the uid/gid mappings, then signals the child to re-exec rpmoci
+
+    let user_id = getuid();
+    let group_id = getgid();
+    // dnf chooses a root-writable cache directory by default. It's unlikely that
+    // the current user will be able to write there, so we configure the RPMOCI_CACHE_DIR
+    // to point to a cache directory in the current user's home directory.
+    let cache_dir = dirs::cache_dir().unwrap().join("rpmoci");
+
+    const STACK_SIZE: usize = 1024 * 1024;
+    let stack: &mut [u8; STACK_SIZE] = &mut [0; STACK_SIZE];
+    // this pipe is for the parent to notify the child when user namespaces mappings
+    // have been configured
+    let (reader, writer) = pipe()?;
+    // create child process with a new user namespace
+    let child = nix::sched::clone(
+        Box::new(|| {
+            // this child process just waits for the parent to notify it before re-execing
+            close(writer).unwrap();
+            read(reader, &mut Vec::new()).unwrap();
+            Command::new(std::env::current_exe().unwrap())
+                .args(std::env::args().skip(1))
+                .env("RPMOCI_CACHE_DIR", cache_dir.clone())
+                .exec();
+            255
+        }),
+        stack,
+        CloneFlags::CLONE_NEWUSER,
+        Some(signal::SIGCHLD as i32),
+    )
+    .context("Clone failed")?;
+
+    // this parent process sets up user namespace mappings, notifies the child to continue,
+    // then waits for the child to exit
+    close(reader)?;
+    let child_proc = PathBuf::from("/proc").join(child.to_string());
+    File::create(child_proc.join("setgroups"))
+        .context("failed to create setgroups file")?
+        .write_all(b"deny")
```

Comment:
You don't need to set this (since you're writing the `gidmap` from the parent namespace) and I guess an RPM's `postinst` _may_ try to move into some subgroups (no idea why, maybe it's coded to join some groups, drop permissions and then do the rest of the `postinst` as an unprivileged user?).  I think "fake root" isn't exempted from `setgroups` policing (where "real root" would be).